### PR TITLE
A68: Random subsetting with rendezvous hashing LB policy

### PR DESCRIPTION
**Pull request by s-matyukevich**
Replaces https://github.com/grpc/proposal/pull/383

Related to https://github.com/grpc/proposal/pull/430  which describes an LB policy that could be used in combination with random subsetting to correct the resulting server-side load imbalance.

### CODE REVIEW

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:31]: I think it would help to define the algorithm in pseudo code. See https://github.com/grpc/proposal/blob/master/A55-xds-stateful-session-affinity.md#lb-policy-for-stateful-session-affinity or https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md#aggregated-connectivity-state for examples of gRFCs that do this.

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:64]: You can use the proto language (this tip was given to me by another maintainer).

```suggestion
```proto
```

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:83]: I think you should replace addresses with endpoints to take [A61](https://github.com/grpc/proposal/blob/master/A61-IPv4-IPv6-dualstack-backends.md) into consideration.

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:23]: I think you need to replace `addresses` with `endpoints` to account for https://github.com/grpc/proposal/blob/master/A61-IPv4-IPv6-dualstack-backends.md, where each endpoint may have multiple addresses.

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:97]: This should be 

```suggestion
```proto
```

textproto is used to define the content of a instance of a protobuf message.

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:32]: You'll have to decide which address, in case the endpoint has more than one (I think you can use the first address?).

File: A68-random-subsetting.md
Comment (atollena) [inline A68-random-subsetting.md:111]: You should probably discuss the trade offs of doing this kind of subsetting in the control plane, since it was discussed in the original proposal.

Comment (s-matyukevich) [reply A68-random-subsetting.md:111]: Yeah, but I posted a link to the tl;dr; of the discussion, so you think this is not enough?

Comment (atollena) [reply A68-random-subsetting.md:111]: I'm thinking of the option of doing random subsetting in the control plane by sending different EDS responses (with different subsets) to each dataplane, or the equivalent with other resolvers. It is simple to implement with xDS and works for Envoy and gRPC. IIRC the main argument for not going that route is the need to have an xDS infrastructure (this is a big barrier for our orgs, and probably others), and existing limitations of https://github.com/envoyproxy/go-control-plane.

This was discussed in https://github.com/grpc/proposal/pull/383/files#r1308024474.

In order to understand this proposal, I think users will need to understand the trade off of doing it as a balancer in each data plane rather than directly in service discovery.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md:33]: updated to use xxhash, This changes the algorithms slightly as we can use random pre-generated seed instead of concatenating salt to each address. The new version is even simpler.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Done.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Done.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Updated the doc to use the first address.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md:31]: Done.

**Comment by s-matyukevich** [comment]
Bump on this. It has been almost a month since the proposal was submitted and no one from gRPC maintainers commented on it yet. cc @markdroth and @ejona86 since you both reviewed previous version of the proposal and have full context.

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:96]: Do we want this field to be optional or required?  If optional, is 20 the right default?

My gut reaction (but I'm certainly open to arguments here) is that we don't really know what default would be appropriate for all environments, so we might want to just make this field required.

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:107]: "Attributes and service config from the old resolver state will be copied to the new one."

I'm not sure what that means.  In general, all of the attributes should be passed down with each resolver update, so I don't think we want to retain attributes from the previous update if they were removed in the new update.

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:125]: I think this will want to be:

```
config.cluster.v3.LoadBalancingPolicy child_policy = 2;
```

Note that it does *not* need to be `repeated`, because in xDS the list of options is inside of the `LoadBalancingPolicy` message.

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:124]: See comments above about whether this is required or optional, what the field type is, etc.

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:4]: You can list me as the approver.

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:80]: This is a very nice property!

File: A68-random-subsetting.md
Comment (markdroth) [inline A68-random-subsetting.md:117]: It's probably worth chatting with @wbpcode to make sure this design would also work for Envoy.  I don't know of any reason it wouldn't, but might make sense to check just in case.

**Review by markdroth — COMMENTED** [review]
Sorry for the delay reviewing this!  Overall, it looks very good -- my comments here are mostly minor.

One of the reasons for the delay was that we were having a bunch of dicussions internally about ways to potentially reduce the number of unnecessary idle connections with policies other than pick_first, and I wanted to make sure that where we landed on that didn't conflict with this proposal.  Just as an FYI, let me sketch out what we have in mind to eventually do for that, so that you know where we're heading.

The general problem that we're trying to solve is that LB policies like RR and WRR proactively maintain connections to all endpoints and therefore waste a lot of resources on idle connections when the client is idle.  To deal with this, we will eventually want to add an LB policy that tracks the number of simultaneous RPCs in flight on the channel over the last N minutes and scale the number of connections accordingly.  This new policy would operate similarly to the subsetting policy described in this gRFC: it would filter the set of addresses passed down to the child policy.

One of the things we were discussing was how this theoretical new policy would interact with the subsetting policy described in this gRFC.  We wound up deciding that we would probably just put this theoretical new policy underneath the subsetting policy, so that the two policies can basically be independent of each other.  Obviously, this will probably make the actual connection distribution a bit less even, but it will do so only in the face of idle clients, and it will not actually break the distribution imposed by the subsetting policy -- it would just eliminate some of the connections that would have been chosen by the subsetting policy but would have been idle.

Anyway, as I mentioned, this gRFC looks very good!  I'd like to get reviews from @ejona86 and @dfawley as well, but I don't foresee any significant blockers.

Please let me know if you have any questions.  Thanks!

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: done.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md:50]: The field should be required, I removed the default value. I moved the config section accordingly to your suggestion, but kept the bullet point for now because It describes the origin of `subset_size` variable used in the pseudocode below.  Can remove it if you still think it is unnecessary.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: done.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Sure, I'll do that.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Updated the value to uint32 and added the comment indicating that value must be >= 1.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Yes, this was a mistake. I removed the default value.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: That is exactly what I meant, but I phrased it poorly, updated to `Attributes and service config from the parent resolver state will be passed down with each resolver update`

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: This was a bad copy-paste. Updated.  
Should the corresponding field in the LB policy config be repeated? I removed repeated keyword for now as in this case the child policy is mandatory.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Changed subset_size type from google.protobuf.UInt32Value to uint32 and made it mandatory. Also copy-pasted corresponding comments from the policy config definition.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Done

File: A68-random-subsetting.md
Comment (markdroth) [reply A68-random-subsetting.md]: The reason for `repeated` here is to allow new policies to be configured without breaking older clients.  The client will iterate through the list of child policies, stopping at the first one that it supports.  So let's say that you start with a bunch of clients that support only policy A, and now you want to move to a new policy B, but not all of your clients support B yet.  The control plane can send the list [B, A], which will allow new clients to use B while still allowing older clients to use A.

In the gRPC service config, the `LoadBalancingConfig` message represents a single policy, so we need to make the field `repeated` wherever it gets used.  In xDS, the `config.cluster.v3.LoadBalancingPolicy` message itself contains [a `repeated` list of policies](https://github.com/envoyproxy/envoy/blob/7f7234be42dd7f90c612c193187a65775c119607/api/envoy/config/cluster/v3/cluster.proto#L1235), so there's no need to make it repeated.  (In hindsight, I wish we'd made the same kind of intermediate message in the gRPC service config, so that we didn't always have to remember to make the field `repeated` wherever we use it.  At the time we were concerned with trying to minimize the length of the JSON representation to allow it to squeeze into DNS TXT records, which is why we chose the current approach.)

So yes, please do add `repeated` in the gRPC service config representation.

**Review by ejona86 — COMMENTED** [review]
> To deal with this, we will eventually want to add an LB policy that tracks the number of simultaneous RPCs in flight on the channel over the last N minutes and scale the number of connections accordingly.

One thing for me was how similar the subsetting would be between the two load balancer policies. But the subsetting algorithm here could be used directly by the other LB policy as well. (The other LB policy could maybe just delegate to this, changing its subset_size over time; or we extend this to support dynamic subset size. All the options remain on the table with this gRFC.)

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md:53]: Replaced with:

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: yes, done.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Added `repeated` back to RandomSubsettingLbConfig.

File: A68-random-subsetting.md
Comment (ejona86) [inline A68-random-subsetting.md:54]: nit: s/previosly/previously/

Would it be improved by adding a comma after "list" in the first sentence?

Comment (s-matyukevich) [reply A68-random-subsetting.md]: fixed.

**Review by ejona86 — APPROVED** [review]

**Comment by atollena** [comment]
> The general problem that we're trying to solve is that LB policies like RR and WRR proactively maintain connections to all endpoints and therefore waste a lot of resources on idle connections when the client is idle. To deal with this, we will eventually want to add an LB policy that tracks the number of simultaneous RPCs in flight on the channel over the last N minutes and scale the number of connections accordingly. This new policy would operate similarly to the subsetting policy described in this gRFC: it would filter the set of addresses passed down to the child policy.

This is also the general problem that we are trying to solve with this gRFC and https://github.com/grpc/proposal/pull/430. And making the subset size dynamic would be ideal, but seems much harder, as the subset size impacts load balancing, and server load must be reasonably predictable. I don't know of any published work on making the number of connections dynamic (no fixed subset size) while also keeping the server load reasonably balanced.

**Comment by ejona86** [comment]
> I don't know of any published work on making the number of connections dynamic (no fixed subset size) while also keeping the server load reasonably balanced.

Basically the theory is you would size a minimum subset like you are doing here to a degree. But if the client causes a lot of load, then you scale up, potentially to all backends. IIRC, we were thinking of defining load as "concurrent RPCs." Most of the discussion was spent avoiding wildly scaling up/down, like with bursty clients, causing damage.

**Review by markdroth — APPROVED** [review]

**Comment by markdroth** [comment]
> This is also the general problem that we are trying to solve with this gRFC and #430. And making the subset size dynamic would be ideal, but seems much harder, as the subset size impacts load balancing, and server load must be reasonably predictable. I don't know of any published work on making the number of connections dynamic (no fixed subset size) while also keeping the server load reasonably balanced. So I'm very curious where you are at with those discussion and if you have rough ideas on how you would address load balancing.

I think the approach that we have in mind would probably result in slightly less ideally balanced server load than the results you've seen, specifically because the connections will no longer be perfectly balanced.  But in cases where the RPC rate from different clients diverge wildly, or where RPC rates are very bursty, maintaining unnecessary idle connections can cost significant amounts of memory and some additional CPU.  I think there are situations where the costs of that unnecessary overhead is expensive enough that it's worth taking steps to reduce it, even if it results in slightly less optimally balanced server load.

Like anything else in this space, it's a bit of a balancing act (pun intended).  There will definitely be cases where the approach that we have in mind won't work well, but there will be cases where it will.  But we think it will be a useful tool to have in our toolbox at some point.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md]: Here is the PR https://github.com/grpc/grpc-proto/pull/157

**Comment by s-matyukevich** [comment]
What are the next steps here? Can someone with the right permissions merge this?

**Comment by ejona86** [comment]
@dfawley, Mark is the listed approver currently, but I'm fine merging this as it seems things are in order. Since the first implementation is in Go, you may want to take a look, but you agree the approvers seems ready?

File: A68-random-subsetting.md
Comment (dfawley) [inline A68-random-subsetting.md:65]: Maybe add `if subset_size >= len(endpoints) { return endpoints }` at the top?

File: A68-random-subsetting.md
Comment (dfawley) [inline A68-random-subsetting.md:136]: Can we somehow guarantee this is always the case?  If not, what happens if envoy adds a field here without telling anyone, e.g.?  Does gRPC only support the fields listed here, and we ignore anything else?

**Comment by dfawley** [comment]
> @dfawley, Mark is the listed approver currently, but I'm fine merging this as it seems things are in order. Since the first implementation is in Go, you may want to take a look, but you agree the approvers seems ready?

This LGTM overall, but I'm a little concerned about this comment from Mark and its impact on this gRFC: https://github.com/grpc/proposal/pull/423/files#r1676227354

Have you had this discussion yet, @s-matyukevich?  Otherwise, should we wait to merge this until that's been finalized (more of a Q for @ejona86)?  @markdroth did approve, so maybe he didn't see it as a blocker.

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md:68]: added

File: A68-random-subsetting.md
Comment (s-matyukevich) [reply A68-random-subsetting.md:139]: Not sure about this. IMO this problem applies to any gRFC that defines matching xds and gRPC LB policies. @wbpcode is currently the [codeowner](https://github.com/envoyproxy/envoy/blob/main/CODEOWNERS#L317-L324) for all xds lb policies so I redirect this question to him.

**Comment by markdroth** [comment]
I don't really anticipate any problems on the Envoy side, but it would be good to get a sanity check from @wbpcode before merging this, just to be safe.  I'll ping him offline to request his input.

File: A68-random-subsetting.md
Comment (markdroth) [reply A68-random-subsetting.md:139]: We can't guarantee that new fields won't be added to xDS (although as an xDS API shepherd, I'd see the change go by if it did), but that's fine -- it's just like any other feature getting implemented in one data plane before another.

The important thing is just that the fields we add to xDS won't get *removed*, and the backward compatibility requirements for xDS guarantee that.

So I think we're fine here.

File: A68-random-subsetting.md
Comment (wbpcode) [reply A68-random-subsetting.md:122]: It is not hard to implement that in envoy. So, it's fine.

**Review by wbpcode — APPROVED** [review]

**Comment by markdroth** [comment]
Thanks for confirmation, @wbpcode!

I think that's all the open questions here, so I'll go ahead and merge.  Thanks!

**Pull request merged**
